db/hybrid_logical_vector.go

```diff
@@ -739,3 +739,126 @@ func IsInConflict(ctx context.Context, localHLV, incomingHLV *HybridLogicalVecto
+		return nil, errors.New("local or incoming document is nil for resolveConflict")
+	}
+	// resolve conflict in favor of remote document, remote wins case
+	if conflict.RemoteHLV.Version > conflict.LocalHLV.Version {
+		// remote document wins
+		return conflict.RemoteDocument, nil
+	}
+	return conflict.LocalDocument, nil
+}
+
+// localWinsConflictResolutionForHLV will alter the HLV for a local wins conflict resolution. Preserving local MV
+// unless incoming MV has a src common with local MV and has a higher version, in which case local MV is invalidated and moved to PV.
+// In the eventuality that local CV is <= to incoming CV, a new CV will be created with this db's sourceID
+func localWinsConflictResolutionForHLV(ctx context.Context, localHLV, incomingHLV *HybridLogicalVector, docID, sourceID string) (*HybridLogicalVector, error) {
+	if localHLV == nil || incomingHLV == nil {
+		return nil, errors.New("local or incoming hlv is nil for resolveConflict")
+	}
+
+	newHLV := localHLV.Copy()
+
+	defer func() {
+		base.DebugfCtx(ctx, base.KeyVV, "resolved conflict for doc %s in favour of local wins, resulting HLV: %v", base.UD(docID), newHLV)
+	}()
+
+	if localHLV.Version > incomingHLV.Version {
+		// no need to generate new CV value
+		newHLV.addNewerVersionTmp(incomingHLV)
+		return newHLV, nil
+	}
+
+	// here local CV version is <= to incoming CV version, thus we need to create a new CV with this db's sourceID then
+	// move any mv to pv
+	base.DebugfCtx(ctx, base.KeyVV, "new cv needed for doc %s in local wins conflict resolution, local hlv: %v, incoming hlv: %v", base.UD(docID), localHLV, incomingHLV)
+	newHLV.SourceID = sourceID
+	newHLV.Version = expandMacroCASValueUint64 // set new version to the macro expansion value
+	newHLV.CurrentVersionCAS = expandMacroCASValueUint64
+
+	newHLV.addNewerVersionsToPV(newHLV.MergeVersions) // add any newer merge version to pv
+	newHLV.MergeVersions = make(HLVVersions)
+	// now add local and incoming docs CV's to the new local HLV merge versions
+	newHLV.MergeVersions[localHLV.SourceID] = localHLV.Version       // add local cv to mv
+	newHLV.MergeVersions[incomingHLV.SourceID] = incomingHLV.Version // add incoming cv to mv
+
+	// now add any newer versions in pv from incoming doc to local docs pv
+	newHLV.addNewerVersionsToPV(incomingHLV.PreviousVersions)
+
+	return newHLV, nil
+}
+
+// remoteWinsConflictResolutionForHLV will alter the HLV for a remote wins conflict resolution. Preserving incoming MV
+// unless local MV has a src common with incoming MV and has a higher version, in which case incoming MV is invalidated and moved to PV.
+func remoteWinsConflictResolutionForHLV(ctx context.Context, docID string, localHLV, incomingHLV *HybridLogicalVector) (*HybridLogicalVector, error) {
+	if localHLV == nil || incomingHLV == nil {
+		return nil, errors.New("local or incoming hlv is nil for resolveConflict")
+	}
+
+	newHLV := incomingHLV.Copy()
+
+	newHLV.addNewerVersionTmp(localHLV)
```

Comment:
I think that the case where:

> Preserving incoming MV unless local MV has a src common with incoming MV and has a higher version, in which case incoming MV is invalidated and moved to PV.

should not be possible as per https://github.com/couchbase/sync_gateway/pull/7687#discussion_r2289443416

Let me know if there's a case you are thinking of that doesn't meet this classification.